### DISCUSSION TITLE
Agentic workflows? Does it plan to be supported?

### DISCUSSION THREAD

**Discussion by tonyppe** [danny-avila/LibreChat#3101] [Feature Requests & Suggestions]
I am trying to understand if this is in the roadmap? 

Agentic workflows = multiple agents that work together, chat together, optionally call tools/plugins, autonomously until task is complete. 

Context: Create a cluster of agents with different persona's eg a project manager, a researcher and a article writer. Pose a question: "What is the most recent news on vaping in Australia" and the outcome is a thoroughly researched article about the topic. 

I had a quick search and can see some others have asked about it here: https://github.com/danny-avila/LibreChat/discussions/2460 . The maintainer's response indicated they may have misunderstood the initial question. While they helpfully pointed to a discussion in #2557, that thread focuses more on extending LibreChat with custom plugins rather than enabling fully agentic workflows. It's possible the maintainer was suggesting that agentic workflows have already been considered but not pursued, can you confirm? The maintainer mentioned another feature would be implemented that copied openai's assistants. But this differs from an agentic workflow. 

An agentic workflow would allow any LLM to be consumed, not solely openai and therefore would be more in line with open source as open source models could be utilised such as meta llama.

**Comment by napter** [comment]
I had the same question - specifically if agents can run autonomously (ie continue through multiple turns/requests like cline)

**Comment by tonyppe** [comment]
napter - lost cause mate. Even if they decide to back-track and implement this now, they have a lot of catch up to do. Shame really.

**Reply by danny-avila** [reply]
Can you elaborate? Agents are implemented. What catch up is there? I've already tested multi-agent workflows as well, and it works well, but I've not implemented since it's highly experimental (and not highly requested).

https://www.librechat.ai/docs/features/agents

An agent can run as many turns as it requires. Long-running for indefinite time is not well suited for an AI chat app, but scheduled tasks similar to ChatGPT is planned.

**Reply by andrearizzini-deciphex** [reply]
Hello, 

not sure if this is the right place to discuss about this, but since we are on the topic, here we go.

I agree with the comments above, i think the way the multi agent is currently implemented in librechat is limited in its capabilities.

For example: 
1. There is no overall orchestration available
2. execution is always sequential
3. There is no recursion / repletion

For example if you have a multi agent defined as follow 

1. Planner Agent  - split task into smaller sub-tasks
2. Execution Assistant  - go and execute
3. Critic Agent - quality check the execution output and list things that are missing
4. Reflection Agent - decision maker, do we need more execution or we can terminate
5. Report Agent - generate a report.

This is what i have implemented in LibreChat
![image](https://github.com/user-attachments/assets/ae82b2a7-4920-4c64-a943-180157d514fd)

Lets say my prompt is (working example): 

`Make a python app to upload a file and save the file name to a SQLite database. You will need to create the database and tables if it doesn’t exist`

So the agents above will resond with following:
1. Planner Agent  - list of tasks
2. Execution Assistant - generate beautifull code
3. Critic Agent - says that is not satistifed, code is missing exception and logging to file
4. Reflection Agent - makes the decision we need to refine the solution
5. Report agent - generates a partial solution

What is essentially missing is the ability to influence the flow. 

If the critic agent is not happy with the result it should be able to pass its critique to the execution assistant which in its turn will incorporate the critique.

so the execution flow would look like 1, 2, 3, 4, 2, 3, 4, 5

The requirement would also need to capture the status of each stage and specific routes.

For example: 

```json
[{
     result: "PASS", 
     call: "reportAgent"
}, {
   result: "FAIL"
   call: "executionAgent"
}]```

**Reply by danny-avila** [reply]
Yea we could do hand offs, then most kinds of workflows could be orchestrated

**Reply by andrearizzini-deciphex** [reply]
and "hand off" you would mean that you only send a query an get back the result? so you will not be able to audit or verify each of the agents interaction?

**Reply by andrearizzini-deciphex** [reply]
handsoff might not be fully auditable, but if the agent returned in the response a specific convention could LibreChat use this to immediatelly submit the query to the other agent? that way we can see what each agent output is and we have full visibility and tracebility..

**Comment by falkenbt** [comment]
Maybe worth looking at the a2a protocol 
* https://developers.googleblog.com/en/a2a-a-new-era-of-agent-interoperability
* https://github.com/google-a2a/A2A

**Comment by Lucas-almma** [comment]
How similar/different is the functionality described as n8n?
They also have an OS repo, not MIT, but very helpful.

ᐧ
